invoices/update_invoice.go

```diff
@@ -0,0 +1,816 @@
+			return &update.NewState, nil, nil
+		}
+
+		// Sanity check that the user isn't trying to settle or accept a
+		// non-existent HTLC set.
+		set := invoice.HTLCSet(update.SetID, HtlcStateAccepted)
+		if len(set) == 0 {
+			return nil, nil, ErrEmptyHTLCSet
+		}
+
+		// For AMP invoices, there are no invoice-level preimage checks.
+		// However, we still sanity check that we aren't trying to
+		// settle an AMP invoice with a preimage.
+		if update.SetID != nil {
+			if update.Preimage != nil {
+				return nil, nil, errors.New("AMP set cannot " +
+					"have preimage")
+			}
+
+			return &update.NewState, nil, nil
+		}
+
+		var preimage *lntypes.Preimage
+
+		switch {
+		// If an invoice-level preimage was supplied, but the InvoiceRef
+		// doesn't specify a hash (e.g. AMP invoices) we fail.
+		case update.Preimage != nil && hash == nil:
+			return nil, preimage,
+				ErrUnexpectedInvoicePreimage
+
+		// Validate the supplied preimage for non-AMP invoices.
+		case update.Preimage != nil:
+			if update.Preimage.Hash() != *hash {
+				return nil, nil,
+					ErrInvoicePreimageMismatch
+			}
+			preimage = update.Preimage
+
+		// Permit non-AMP invoices to be accepted without knowing the
+		// preimage. When trying to settle we'll have to pass through
+		// the above check in order to not hit the one below.
+		case update.NewState == ContractAccepted:
+
+		// Fail if we still don't have a preimage when transitioning to
+		// settle the non-AMP invoice.
+		case update.NewState == ContractSettled &&
+			invoice.Terms.PaymentPreimage == nil:
+
+			return nil, nil, errors.New("unknown preimage")
+		}
+
+		return &update.NewState, preimage, nil
+
+	// Once settled, we are in a terminal state.
+	case ContractSettled:
+		return nil, nil, ErrInvoiceAlreadySettled
+
+	// Once canceled, we are in a terminal state.
+	case ContractCanceled:
+		return nil, nil, ErrInvoiceAlreadyCanceled
+
+	default:
+		return nil, nil, errors.New("unknown state transition")
+	}
+}
+
+// canCancelSingleHtlc validates cancellation of a single HTLC. If nil is
+// returned, then the HTLC can be cancelled.
+func canCancelSingleHtlc(htlc *InvoiceHTLC,
+	invoiceState ContractState) error {
+
+	// It is only possible to cancel individual htlcs on an open invoice.
+	if invoiceState != ContractOpen {
+		return fmt.Errorf("htlc canceled on invoice in state %v",
+			invoiceState)
+	}
+
+	// It is only possible if the htlc is still pending.
+	if htlc.State != HtlcStateAccepted {
+		return fmt.Errorf("htlc canceled in state %v", htlc.State)
+	}
+
+	return nil
+}
+
+// getUpdatedHtlcState aligns the state of an htlc with the given invoice state.
+// A boolean indicating whether the HTLCs state need to be updated, along with
+// the new state (or old state if no change is needed) is returned.
+func getUpdatedHtlcState(htlc *InvoiceHTLC,
+	invoiceState ContractState, setID *[32]byte) (
+	bool, HtlcState, error) {
+
+	trySettle := func(persist bool) (bool, HtlcState, error) {
```

Comment:
Since this commit is just moving code around, perhaps best to recreate these comments on other commits as every time I rebase I have to manually recreate this commit. Also maybe just to reduce the scope of the PR, perhaps we could do these refactors in a follow up. WDYT?